irohad/torii/impl/command_service.cpp

```diff
@@ -130,4 +136,126 @@ namespace torii {
+    if (not finished) {
+      if (not resp) {
+        subscription.unsubscribe();
+        iroha::protocol::ToriiResponse resp_none;
+        resp_none.set_tx_hash(request_hash);
+        resp_none.set_tx_status(iroha::protocol::TxStatus::NOT_RECEIVED);
+        response_writer.WriteLast(resp_none, grpc::WriteOptions());
+      } else {
+        // Tx processing was started but still unfinished. We give it more time.
+        cv.wait_for(lock, finish_tx_processing_duration_);
+      }
+    }
+  }
+
+  grpc::Status CommandService::StatusStream(
+      grpc::ServerContext *context,
+      const iroha::protocol::TxStatusRequest *request,
+      grpc::ServerWriter<iroha::protocol::ToriiResponse> *response_writer) {
+    StatusStream(*request, *response_writer);
+    return grpc::Status::OK;
+  }
+
+  void CommandService::checkCacheAndSend(
+      boost::optional<iroha::protocol::ToriiResponse> &resp,
+      grpc::ServerWriter<iroha::protocol::ToriiResponse> &response_writer)
+      const {
+    if (resp) {
+      if (resp->tx_status()
+              == iroha::protocol::TxStatus::STATELESS_VALIDATION_FAILED
+          or resp->tx_status()
+              == iroha::protocol::TxStatus::STATEFUL_VALIDATION_FAILED
+          or resp->tx_status() == iroha::protocol::TxStatus::NOT_RECEIVED
+          or resp->tx_status() == iroha::protocol::TxStatus::COMMITTED) {
+        response_writer.WriteLast(*resp, grpc::WriteOptions());
+        return;
+      }
+      response_writer.Write(*resp);
+    }
+  }
+
+  iroha::protocol::TxStatus CommandService::convertStatusToProto(
+      iroha::model::TransactionResponse::Status &status) {
+    iroha::protocol::TxStatus proto_status;
+    switch (status) {
+      case iroha::model::TransactionResponse::STATELESS_VALIDATION_FAILED:
+        proto_status = iroha::protocol::TxStatus::STATELESS_VALIDATION_FAILED;
+        break;
+      case iroha::model::TransactionResponse::STATELESS_VALIDATION_SUCCESS:
+        proto_status = iroha::protocol::TxStatus::STATELESS_VALIDATION_SUCCESS;
+        break;
+      case iroha::model::TransactionResponse::STATEFUL_VALIDATION_FAILED:
+        proto_status = iroha::protocol::TxStatus::STATEFUL_VALIDATION_FAILED;
+        break;
+      case iroha::model::TransactionResponse::STATEFUL_VALIDATION_SUCCESS:
+        proto_status = iroha::protocol::TxStatus::STATEFUL_VALIDATION_SUCCESS;
+        break;
+      case iroha::model::TransactionResponse::IN_PROGRESS:
+        proto_status = iroha::protocol::TxStatus::IN_PROGRESS;
+        break;
+      case iroha::model::TransactionResponse::COMMITTED:
+        proto_status = iroha::protocol::TxStatus::COMMITTED;
+        break;
+      case iroha::model::TransactionResponse::NOT_RECEIVED:
+      default:
+        proto_status = iroha::protocol::TxStatus::NOT_RECEIVED;
+        break;
+    }
+    return proto_status;
+  }
+
+  bool CommandService::isFinalStatus(iroha::protocol::TxStatus &status) {
```

Comment:
Maybe there is more elegant way to do it?

Reply:
Well, @muratovv advised something similar in one of his comments. I think introduction a new collection of final statuses will just make code larger. If we decide to introduce one more final status both approaches will require a change anyway.